Common/include/toolboxes/graph_toolbox.hpp

```diff
@@ -0,0 +1,489 @@
+    innerIdx.insert(innerIdx.end(), neighbors.begin(), neighbors.end());
+  }
+  outerPtr.back() = innerIdx.size();
+
+  /*--- Return pattern as CCompressedSparsePattern object. ---*/
+  return CCompressedSparsePattern<Index_t>(outerPtr, innerIdx);
+}
+
+
+/*!
+ * \brief Build a lookup table of the absolute positions of the non zero entries
+ *        of a compressed sparse pattern, accessed when visiting the FVM edges
+ *        of a grid. The table can then be used for fast access (avoids searches)
+ *        to the non zero entries of a sparse matrix associated with the pattern.
+ * \param[in] geometry - Definition of the grid.
+ * \param[in] pattern - Sparse pattern.
+ * \return nEdge by 2 matrix.
+ */
+template<class Geometry_t, typename Index_t>
+CEdgeToNonZeroMap<Index_t> mapEdgesToSparsePattern(Geometry_t& geometry,
+                                                   const CCompressedSparsePattern<Index_t>& pattern)
+{
+  assert(!pattern.empty());
+
+  CEdgeToNonZeroMap<Index_t> edgeMap(geometry.GetnEdge(),2);
+
+  for(Index_t iEdge = 0; iEdge < geometry.GetnEdge(); ++iEdge)
+  {
+    Index_t iPoint = geometry.edge[iEdge]->GetNode(0);
+    Index_t jPoint = geometry.edge[iEdge]->GetNode(1);
+
+    edgeMap(iEdge,0) = pattern.quickFindInnerIdx(iPoint,jPoint);
+    edgeMap(iEdge,1) = pattern.quickFindInnerIdx(jPoint,iPoint);
+  }
+
+  return edgeMap;
+}
+
+
+/*!
+ * \brief Color contiguous groups of outer indices of a sparse pattern such that
+ *        within each color, any two groups do not have inner indices in common.
+ * \note  Within a group, two outer indices will generally have common inner indices.
+ *        The coloring is returned as a compressed sparse pattern where the colors
+ *        are outer indices, and the outer indices of the input pattern are the
+ *        inner indices of the coloring. A simple greedy algorithm is used.
+ *        Using a sparse pattern as input allows "anything" to be colored e.g.
+ *        FVM edges, FEM elements, the rows/columns of a sparse matrix, etc.
+ * \note  The worst that can happen in this method is needing an unreasonable number
+ *        of colors, or too much memory due to a large range of the inner indices.
+ *        The last two template parameters limit both, in case of failure an empty
+ *        pattern is returned.
+ * \param[in] pattern - Sparse pattern to be colored.
+ * \param[in] groupSize - Size of the outer index groups, default 1.
+ * \param[out] indexColor - Optional, vector with colors given to the outer indices.
+ * \return Coloring in the same type of the input pattern.
+ */
+template<class T, typename Color_t = char, size_t MaxColors = 64, size_t MaxMB = 128>
+T colorSparsePattern(const T& pattern, size_t groupSize = 1,
+                     std::vector<Color_t>* indexColor = nullptr)
+{
+  static_assert(std::is_integral<Color_t>::value,"");
+  static_assert(std::numeric_limits<Color_t>::max() >= MaxColors,"");
+
+  using Index_t = typename T::IndexType;
```

Comment:
Greedy coloring of one of those sparse patterns, I did it like this to avoid having to implement one version for edges, one for elements, etc. I know we had some coloring stuff laying around but this only needs to operate within the rank, and it supports coloring groups instead of single entities of the sparse pattern (as shown in #789 that will be quite important for performance).